platform/kube/admit/admit.go

```diff
@@ -0,0 +1,248 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package admit
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/ghodss/yaml"
+	"github.com/golang/glog"
+	"k8s.io/api/admission/v1alpha1"
+	admissionregistrationv1alpha1 "k8s.io/api/admissionregistration/v1alpha1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/client-go/kubernetes"
+	admissionClient "k8s.io/client-go/kubernetes/typed/admissionregistration/v1alpha1"
+
+	"istio.io/pilot/adapter/config/crd"
+	"istio.io/pilot/model"
+)
+
+// A cluster-unique (i.e. random) suffix should be added to each
+// default below when testing in a shared cluster to avoid collisions.
+
+const (
+	// DefaultAdmissionHookConfigName is the default name for the
+	// ExternalAdmissionHookConfiguration.
+	DefaultAdmissionHookConfigName = "pilot-config"
+
+	// DefaultAdmissionHookName is the default name for the
+	// ExternalAdmissionHooks.
+	DefaultAdmissionHookName = "pilot.config.istio.io"
+
+	// DefaultAdmissionServiceName is the default service of the
+	// validation webhook.
+	DefaultAdmissionServiceName = "istio-pilot-config"
+)
+
+// Admit implements the external admission webhook for validation
+// pilot configuration.
+type Admit struct {
+	descriptor       model.ConfigDescriptor
+	hookConfigName   string
+	hookName         string
+	serviceName      string
+	serviceNamespace string
+	caBundle         []byte
+
+	// unconditionally validate all config that is not in this list of
+	// configuration.
+	validateNamespaces []string
+}
+
+// GetAPIServerExtensionCACert gets the CA cert that will signed the
+// cert used by the "GenericAdmissionWebhook" plugin admission
+// controller.
+func GetAPIServerExtensionCACert(cl kubernetes.Interface) ([]byte, error) {
+	const name = "extension-apiserver-authentication"
+	c, err := cl.CoreV1().ConfigMaps(metav1.NamespaceSystem).Get(name, metav1.GetOptions{})
```

Comment:
Why not a secret to store CA cert?

Reply:
Is kube-system the right place to store the config? Is the name of the config standard?

Reply:
That is where Kubernetes stores the aggregate apiserver's client CA certificate. We don't control this.

```bash
$ kubectl -n kube-system get cm extension-apiserver-authentication 
NAME                                 DATA      AGE
extension-apiserver-authentication   6         17d
```

Reply:
Please mention this. It might appear as if we're choosing these names.